src/ripple/overlay/impl/MessageScheduler.cpp

```diff
@@ -0,0 +1,413 @@
+#include <ripple/basics/Log.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/overlay/MessageScheduler.h>
+
+#include <cassert>
+#include <chrono>
+#include <exception>
+#include <iterator>
+#include <mutex>
+
+namespace ripple {
+
+using RequestId = MessageScheduler::RequestId;
+
+// We use a thread-local variable to track whether a sender is being
+// scheduled from a callback in a thread where the scheduler is already
+// locked. In that case, trying to lock the scheduler again is undefined
+// behavior. Instead, we just save the sender here to be handled after the
+// callback returns.
+thread_local static std::vector<MessageScheduler::Sender*>* tsenders = nullptr;
```

Comment:
Instead of a thread local consider re-designing things so locks are not held during callbacks and new messages can be scheduled from callbacks without special logic. One way to do that is to do things in two passes - once under a lock where you might copy some things into another local collection, and once without the lock where you iterate though the local collection. It would be a win if we could re-design this so we remove this thread-local as well as not holding locks during callbacks.